lib/improver/utilities/indexing_operations.py

```diff
@@ -0,0 +1,168 @@
+
+import numpy as np
+
+
+def choose(index_array, array_set):
+    """
+    Construct an array from an index array and a set of arrays to choose from.
+
+    The index values specify which array to take the value from. For example,
+    if the function is provided with a list of 10 arrays, the index values
+    should range from 0-9. This function should be provided with a
+    multi-dimensional array and is not designed to accept a list of arrays,
+    which is the preferred input to numpy choose. The first dimension of the
+    multi-dimensional array will be treated as defining the sequence of
+    sub-arrays.
+
+    The shape of the index array, and where in that shape the array indices
+    are positioned will determine which data is extracted from the arrays. The
+    two arrays will be broadcast to a common shape if they do not match and
+    such a broadcasting operation is possible. If the input index_array shape
+    does not match the array_set shape, it is assumed that the indexing is for
+    the sub-arrays, not the complete array_set. The leading dimension in the
+    resulting extracted array is thus cut away such that the returned array
+    will have the same shape as the sub-arrays. If the input index_array has
+    the same number of dimensions as the array_set, the returned array will
+    have an equivalent number of dimensions (i.e. one more than the
+    sub-arrays).
+
+    The following figure gives some examples of the expected result of this
+    function when a series of index_arrays of differing shape and
+    dimensionality are used. The array_set is the same for each example,
+    shown down the left hand side; this array has the shape (3, 2, 2).
+
+    .. figure:: extended_documentation/utilities/indexing_operations/
+       numpy_choose_example.png
+       :scale: 60 %
+       :alt: Diagram to help explain indexing behaviour of choose.
+
+       The Array column gives the index in the outer most dimension of the
+       multi-dimensional input array (array_set). The Data column shows the
+       contents of the 3 sub-arrays that correspond to the given Array index.
+       The Indices examples are used to show which data would be returned
+       (Result) should the index_array be shaped in these various ways, with
+       these array index values. The colour coding is used to help make clear
+       how the shape of the index_array relates to the arrays within the
+       array_set.
+
+    The examples above demonstrate the application of choose with a series of
+    index_arrays of different dimensionality. In all but the last of these
+    cases the choose function will use broadcasting to construct an index array
+    that matches the shape of the sub-arrays from which data is being taken.
+    The figure below illustrates the results of this intermediate broadcasting
+    step.
+
+    .. figure:: extended_documentation/utilities/indexing_operations/
+       numpy_choose_broadcasting.png
```

Comment:
This png is a nice illustration but took me a while to decode.  Is there a way to add eg a vertical line between the data array, and the different index behaviours?  Otherwise it looks like a single table where the rows don't make any sense.

Reply:
Done.